### PR TITLE
[Console] Don't go past exact matches when autocompleting

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #21789
| License       | MIT
| Doc PR        | -

### CODE REVIEW

File: src/Symfony/Component/Console/Tests/Helper/QuestionHelperTest.php [hunk 1]
```diff
@@ -160,6 +160,31 @@ public function testAskWithAutocompleteWithNonSequentialKeys()
         $this->assertEquals('AsseticBundle', $dialog->ask($this->createInputInterfaceMock(), $this->createOutputInterface(), $question));
     }
 
+    public function testAskWithAutocompleteWithExactMatch()
+    {
+        if (!$this->hasSttyAvailable()) {
+            $this->markTestSkipped('`stty` is required to test autocomplete functionality');
+        }
+
+        // <UP ARROW><UP ARROW><NEWLINE><DOWN ARROW><DOWN ARROW><NEWLINE>
```
Comment (ogizanagi) [inline src/Symfony/Component/Console/Tests/Helper/QuestionHelperTest.php]: Looks like just "b" + newline to me :)

**Review by ogizanagi — APPROVED** [review]

File: src/Symfony/Component/Console/Tests/Helper/QuestionHelperTest.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Console/Tests/Helper/QuestionHelperTest.php]: fixed :)

File: src/Symfony/Component/Console/Tests/Helper/QuestionHelperTest.php [hunk 2]
```diff
@@ -160,6 +160,30 @@ public function testAskWithAutocompleteWithNonSequentialKeys()
...
+        );
+
+        $dialog = new QuestionHelper();
+        $dialog->setInputStream($inputStream);
+        $dialog->setHelperSet(new HelperSet(array(new FormatterHelper())));
+
+        $question = new ChoiceQuestion('Please select a city', $possibleChoices);
+        $question->setMaxAttempts(1);
+
+        $this->assertEquals('b', $dialog->ask($this->createInputInterfaceMock(), $this->createOutputInterface(), $question));
```
Comment (phansys) [inline src/Symfony/Component/Console/Tests/Helper/QuestionHelperTest.php]: Could `assertSame()` be used instead?

**Review by phansys — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Console/Tests/Helper/QuestionHelperTest.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Console/Tests/Helper/QuestionHelperTest.php]: fixed

**Review by chalasr — APPROVED** [review]

File: src/Symfony/Component/Console/Tests/Helper/QuestionHelperTest.php [hunk 3]
```diff
@@ -160,6 +160,30 @@ public function testAskWithAutocompleteWithNonSequentialKeys()
...
+        );
+
+        $dialog = new QuestionHelper();
+        $dialog->setInputStream($inputStream);
+        $dialog->setHelperSet(new HelperSet(array(new FormatterHelper())));
+
+        $question = new ChoiceQuestion('Please select a city', $possibleChoices);
+        $question->setMaxAttempts(1);
+
+        $this->assertSame('b', $dialog->ask($this->createInputInterfaceMock(), $this->createOutputInterface(), $question));
```
Comment (chalasr) [inline src/Symfony/Component/Console/Tests/Helper/QuestionHelperTest.php]: for merger: `getInputStream()` to be replaced by `createStreamableInputInterfaceMock()` on 3.4

**Comment by chalasr** [comment]
Thank you @nicolas-grekas.

**Comment by lcp0578** [comment]
on windows MINGW32.
>  php bin/console kit:doctrine:generate:entity -vvv
> 
>   Welcome to the kitlabs Doctrine2 entity generator
> 
> This command helps you generate Doctrine2 entities.
> 
> First, you need to give the entity name you want to generate.
> You must use the shortcut notation like AcmeBlogBundle:Post.
> 
> The Entity shortcut name: ProcessBundle:Test
> P[K7rocessBundle8r[K7ocessBundle8o[K7cessBundle8c[K7essBundle8e[K7ssBundle8s[K7sBundle8s[K7Bundle8B[K7undle8u[K7ndle8n[K7dle8d[K7le8l[K7e
> 8e[K78:[KT[Ke[Ks[Kt[K
> Enter table comment: 测试
> 
> In QuestionHelper.php line 175:
> 
>   [Symfony\Component\Console\Exception\RuntimeException]
>   Aborted.
> 
> 
> Exception trace:
>  () at D:\phpStudy\WWW\space2019\law\vendor\symfony\symfony\src\Symfony\Component\Console\Helper\QuestionHelper.php:175
>  Symfony\Component\Console\Helper\QuestionHelper->doAsk() at D:\phpStudy\WWW\space2019\law\vendor\symfony\symfony\src\Symfony\Component\Console\Helper\QuestionHelper.php:80
> 
>  Symfony\Component\Console\Helper\QuestionHelper->ask() at D:\phpStudy\WWW\space2019\law\vendor\kitlabs\kit-generator-bundle\Command\GenerateDoctrineEntityCommand.php:167
>  Kit\GeneratorBundle\Command\GenerateDoctrineEntityCommand->interact() at D:\phpStudy\WWW\space2019\law\vendor\symfony\symfony\src\Symfony\Component\Console\Command\Command
> .php:240
>  Symfony\Component\Console\Command\Command->run() at D:\phpStudy\WWW\space2019\law\vendor\symfony\symfony\src\Symfony\Component\Console\Application.php:987
>  Symfony\Component\Console\Application->doRunCommand() at D:\phpStudy\WWW\space2019\law\vendor\symfony\symfony\src\Symfony\Bundle\FrameworkBundle\Console\Application.php:86
> 
>  Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() at D:\phpStudy\WWW\space2019\law\vendor\symfony\symfony\src\Symfony\Component\Console\Application.php:25
> 5
>  Symfony\Component\Console\Application->doRun() at D:\phpStudy\WWW\space2019\law\vendor\symfony\symfony\src\Symfony\Bundle\FrameworkBundle\Console\Application.php:74
>  Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at D:\phpStudy\WWW\space2019\law\vendor\symfony\symfony\src\Symfony\Component\Console\Application.php:148
>  Symfony\Component\Console\Application->run() at D:\phpStudy\WWW\space2019\law\bin\console:27
> 
> kit:doctrine:generate:entity [--entity [ENTITY]] [--table_comment [TABLE_COMMENT]] [--fields FIELDS] [--format FORMAT] [-h|--help] [-q|--quiet] [-v|vv|vvv|--verbose] [-V|--
> version] [--ansi] [--no-ansi] [-n|--no-interaction] [-e|--env ENV] [--no-debug] [--] <command> [<entity>]


the code: \vendor\symfony\symfony\src\Symfony\Component\Console\Helper\QuestionHelper.php:175
 ``` php
  if (false === $ret) {
                $ret = fgets($inputStream, 4096);
                if (false === $ret) {
                    throw new RuntimeException('Aborted.');
                }
                $ret = trim($ret);
            }
```
but, when i use cmd.exe i can't input chinese word.
on my macbook it it ok.
my extends bundle [github.com/kitlabs-cn/KitGeneratorBundle](https://github.com/kitlabs-cn/KitGeneratorBundle)

**Comment by lcp0578** [comment]
#31870

**Pull request merged**